File: js/bin/keyschedule_check.js
// Reads the shared key-schedule vector
// (spec/corpus/transcripts/keyschedule.json) and confirms this JS symmetric
// state reproduces every output. Agreement with Java, Elixir, Rust, and Go over
// this file means a handshake driven by any of them derives the same transport
// keys. Invoked by interop/check-keyschedule-js.sh.
import fs from 'node:fs';
import { KeySchedule } from '../src/keyschedule.js';

const path = process.argv[2];
if (!path) { process.stderr.write('usage: keyschedule_check <keyschedule.json>\n'); process.exit(2); }
const { inputs, outputs } = JSON.parse(fs.readFileSync(path, 'utf8'));
const hex = (h) => Buffer.from(h, 'hex');

let failures = 0;
const check = (name, want, got) => {
  if (got.toString('hex') === want) { console.log(`PASS ${name}`); }
  else { console.log(`FAIL ${name}\n  got:  ${got.toString('hex')}\n  want: ${want}`); failures++; }
};

const s = new KeySchedule();
check('h_init', outputs.h_init, s.h);
s.mixHash(hex(inputs.mesh_hex));
check('h_after_mesh', outputs.h_after_mesh, s.h);
s.mixKey(hex(inputs.ss_dh_hex));
check('ck_after_dh', outputs.ck_after_dh, s.ck);
s.mixKey(hex(inputs.ss_kem_hex));
check('ck_after_kem', outputs.ck_after_kem, s.ck);
check('ct1', outputs.ct1_hex, s.encryptAndHash(hex(inputs.plaintext1_hex)));
check('h_after_ct1', outputs.h_after_ct1, s.h);
check('ct2', outputs.ct2_hex, s.encryptAndHash(hex(inputs.plaintext2_hex)));
check('h_after_ct2', outputs.h_after_ct2, s.h);
const [i2r, r2i] = s.split();
check('transport_key_i2r', outputs.transport_key_i2r, i2r);
check('transport_key_r2i', outputs.transport_key_r2i, r2i);

if (failures > 0) { process.stderr.write(`${failures} output(s) mismatched\n`); process.exit(1); }
console.log('key schedule reproduces every shared output');

File: rust/src/bin/keyschedule_check.rs
//! Cross-language interop check: reproduces the shared key-schedule vector
//! (spec/corpus/transcripts/keyschedule.json) with the Rust implementation.
//! The Java, Go, and Elixir runners reproduce the same vector, so matching
//! every output means all four derive identical transport keys.

use std::process::exit;

use bonemesh::keyschedule::KeySchedule;
use serde_json::Value;

fn hex_decode(s: &str) -> Vec<u8> {
    (0..s.len()).step_by(2).map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap()).collect()
}

fn hex_encode(b: &[u8]) -> String {
    b.iter().map(|x| format!("{x:02x}")).collect()
}

fn main() {
    let path = std::env::args().nth(1).expect("path to keyschedule.json");
    let doc: Value = serde_json::from_str(&std::fs::read_to_string(&path).unwrap()).unwrap();
    let i = &doc["inputs"];
    let o = &doc["outputs"];
    let mut fails = 0;

    let mut check = |label: &str, got: &[u8], want: &str| {
        if hex_encode(got) == want {
            println!("PASS {label}");
        } else {
            println!("FAIL {label}\n  got:  {}\n  want: {want}", hex_encode(got));
            fails += 1;
        }
    };

    let mut s = KeySchedule::new();
    check("h_init", &s.h, o["h_init"].as_str().unwrap());
    s.mix_hash(&hex_decode(i["mesh_hex"].as_str().unwrap()));
    check("h_after_mesh", &s.h, o["h_after_mesh"].as_str().unwrap());
    s.mix_key(&hex_decode(i["ss_dh_hex"].as_str().unwrap()));
    check("ck_after_dh", &s.ck, o["ck_after_dh"].as_str().unwrap());
    s.mix_key(&hex_decode(i["ss_kem_hex"].as_str().unwrap()));
    check("ck_after_kem", &s.ck, o["ck_after_kem"].as_str().unwrap());
    let ct1 = s.encrypt_and_hash(&hex_decode(i["plaintext1_hex"].as_str().unwrap()));
    check("ct1", &ct1, o["ct1_hex"].as_str().unwrap());
    check("h_after_ct1", &s.h, o["h_after_ct1"].as_str().unwrap());
    let ct2 = s.encrypt_and_hash(&hex_decode(i["plaintext2_hex"].as_str().unwrap()));
    check("ct2", &ct2, o["ct2_hex"].as_str().unwrap());
    check("h_after_ct2", &s.h, o["h_after_ct2"].as_str().unwrap());
    let (i2r, r2i) = s.split();
    check("transport_key_i2r", &i2r, o["transport_key_i2r"].as_str().unwrap());
    check("transport_key_r2i", &r2i, o["transport_key_r2i"].as_str().unwrap());

    drop(check);
    if fails > 0 {
        eprintln!("{fails} key-schedule output(s) mismatched");
        exit(1);
    }
    println!("all key-schedule outputs match");
}
